x/poolmanager/keeper.go

```diff
@@ -208,3 +221,65 @@ func (k *Keeper) SetStakingKeeper(stakingKeeper types.StakingKeeper) {
 func (k *Keeper) SetProtorevKeeper(protorevKeeper types.ProtorevKeeper) {
 	k.protorevKeeper = protorevKeeper
 }
+
+// SetWasmKeeper sets wasm keeper
+func (k *Keeper) SetWasmKeeper(wasmKeeper types.WasmKeeper) {
+	k.wasmKeeper = wasmKeeper
+}
+
+// BeginBlock sets the poolmanager caches if they are empty
+func (k *Keeper) BeginBlock(ctx sdk.Context) {
+	// Set the caches if they are empty.
+	if len(k.cachedTakerFeeShareAgreementMap) == 0 || len(k.cachedRegisteredAlloyPoolByAlloyDenomMap) == 0 || len(k.cachedRegisteredAlloyedPoolIdArray) == 0 {
+		err := k.setTakerFeeShareAgreementsMapCached(ctx)
+		if err != nil {
+			ctx.Logger().Error(fmt.Errorf("%w", types.ErrSetTakerFeeShareAgreementsMapCached).Error())
+		}
+		err = k.setAllRegisteredAlloyedPoolsByDenomCached(ctx)
+		if err != nil {
+			ctx.Logger().Error(fmt.Errorf("%w", types.ErrSetAllRegisteredAlloyedPoolsByDenomCached).Error())
+		}
+		err = k.setAllRegisteredAlloyedPoolIdArrayCached(ctx)
+		if err != nil {
+			ctx.Logger().Error(fmt.Errorf("%w", types.ErrSetAllRegisteredAlloyedPoolsIdArrayCached).Error())
+		}
+	}
+}
+
+// AlloyedAssetCompositionUpdateRate is the rate in blocks at which the taker fee share alloy composition is updated in the end block.
+var AlloyedAssetCompositionUpdateRate = int64(700)
+
+// EndBlock updates the taker fee share alloy composition for all registered alloyed pools
+// if the current block height is a multiple of the alloyedAssetCompositionUpdateRate.
+func (k *Keeper) EndBlock(ctx sdk.Context) {
+	if ctx.BlockHeight()%AlloyedAssetCompositionUpdateRate == 0 {
+		for _, id := range k.cachedRegisteredAlloyedPoolIdArray {
+			err := k.recalculateAndSetTakerFeeShareAlloyComposition(ctx, id)
+			if err != nil {
+				ctx.Logger().Error(fmt.Errorf(
+					"%s for pool id %d: %v", types.ErrSetRegisteredAlloyedPool, id, err,
+				).Error())
+			}
+		}
+	}
+}
+
+// getCacheTrackers returns the cached trackers, used for testing.
+// nolint: unused
+func (k Keeper) getCacheTrackers() (map[string]types.TakerFeeShareAgreement, map[string]types.AlloyContractTakerFeeShareState, []uint64) {
+	return k.cachedTakerFeeShareAgreementMap, k.cachedRegisteredAlloyPoolByAlloyDenomMap, k.cachedRegisteredAlloyedPoolIdArray
+}
```

Comment:
nit: let's define this in the `export_test.go` file

Reply:
https://github.com/osmosis-labs/osmosis/pull/8310/commits/a636a8972c541f16e1087ed9ed8847751979c773